src/overlay/Peer.cpp

```diff
@@ -42,6 +43,133 @@
 namespace stellar
 {
 
+// When flow control is enabled, peer asks for more data every time it processes
+// capacity / FLOW_CONTROL_SEND_MORE_PIPELINE_RATIO messages
+const uint32_t FLOW_CONTROL_SEND_MORE_PIPELINE_RATIO = 4;
```

Comment:
chatted offline about this. Relying completely on TCP sliding window logic would require us to have multiple connections for different traffic types. This is because in order to process SCP messages completely, we need to fetch data from the network. So if we pause reading, we'd never complete fetching and free up capacity, creating a deadlock. 

In this implementation, we're handling two streams of data already ("flood" and "control" traffic), where control traffic is throttled via throttled reading, and flood traffic is throttled by explicitly communicating the load status to the sender (the latter also gives us some flexibility, meaning that in the future we could attach some relevant info for the sender to make more intelligent decision on what to flood next)

Reply:
Have you considered using the QUIC protocol? It supports multiple streams per connection, and each stream has its own, 
independent flow control. So you could have one stream per queue on a single connection.

Reply:
re: window size (re-sharing for posterity)
main thing is to make sure that the pipelining is sufficient, which is directly related to window size. meaning: if the receiver just processed a batch of size B, and asked for the next batch, it should have enough items left to process while it’s waiting for the next batch.

I'm thinking of the worst case scenario: when the sender always has something to send (i.e. it’s producing outbound traffic at a fast rate) and the connection latency is high. I was thinking that 1 round-trip time (outbound refill request, and inbound flood messages) should roughly be equal to `(window size - batch size) * average time it takes to process a task`. I think round trip is something like 30ms-400ms (i.e. nodes within the same state, vs inter-continental). 

I think we can set the window size based on the upper bounds: 400 ms latency, and sender producing messages at 1000 msgs/s (or 1 msg/ms). So setting the window size to something between 500 and 1000 should account for a conservative scenario. When I experimented with tier1 mission and window size 500, the metrics were very similar, and I didn’t see a significant drop in read rate and stuff like that (that's just one specific setup though)

Reply:
We discussed this in chat and I agree with this arithmetic. Marking resolved.